python/lsst/ci/build.py

```diff
@@ -345,6 +395,148 @@ def run(args):
+            b.write_status()
+
+            # Final status posting based on retcode
+            if retcode:
+                if pr_info:
+                    post_statuses.post_all_status(
+                        pr_info, agent=agent, state="success", description=f"Build succeeded on {agent}"
+                    )
+                else:
+                    print(f"Build succeeded on {agent}.")
+
+                sys.exit(0)
+            else:
+                if pr_info:
+                    post_statuses.post_all_status(
+                        pr_info, agent=agent, state="failure", description=f"Build failed on {agent}"
+                    )
+                else:
+                    print(f"Build failed on {agent}")
+                sys.exit(1)
+
+
+class UnknownAgentError(Exception):
+    """Raised when the agent is unknown.
+
+    Parameters
+    ----------
+    agent : `str`
+        The agent that could not be determined.
+    """
+
+    def __init__(self, agent: str) -> None:
+        super().__init__(f"Unknown agent '{agent}', ensure sys.platform is available. Continuing build.")
+        self.agent = agent
+
+
+class PostStatuses:
+    """Class that posts Github Statuses."""
+
+    def post_github_status(self, pr_info: GitHubPR, state: str, description: str, agent: str) -> None:
+        """Post a *single* GitHub status to `pr_info` commit.
+
+        Parameters
+        ----------
+        pr_info : `GitHubPR`
+            The pull request via pydantic model and commit for status.
+        state : `str`
+            The state of the Github status.
+        description : `str`
+            Description of the current status.
+        agent : `str`
+            Current agent that is running.
+        """
+        token = os.getenv("GITHUB_TOKEN")
+        if not token:
+            print("GITHUB_TOKEN not found in environment variables.")
+            return
+
+        owner = pr_info.owner
+        repo = pr_info.repo
+        sha = pr_info.sha
+
+        url = f"https://api.github.com/repos/{owner}/{repo}/statuses/{sha}"
+        headers = {
+            "Authorization": f"token {token}",
+            "Accept": "application/vnd.github.v3+json",
+        }
+
+        print(f"Posting GitHub status to ({owner}/{repo}): {state} - {description}")
+
+        build_url = os.getenv(
+            "RUN_DISPLAY_URL",
+            f"{os.getenv('JENKINS_URL', 'https://rubin-ci.slac.stanford.edu')}"
+            "/blue/organizations/jenkins/stack-os-matrix/activity",
+        )
+
+        data = {
+            "state": state,
+            "description": description,
+            "context": f"Jenkins Build ({agent})",
+            "target_url": build_url,
+        }
+
+        response = requests.post(url, headers=headers, json=data, timeout=120)
+        if response.status_code != 201:
+            print(f"Failed to post GitHub status: {response.status_code} - {response.text}")
+
+    def post_all_status(
```

Comment:
Consider making this a method on `AllProductPRCollections` and the above a method on `GitHubPR`. The PostStatuses class doesn't have any state and these both look like methods on the state of the associated classes.